components/x-live-blog-wrapper/src/dispatchEvent.js

```diff
@@ -0,0 +1,34 @@
+/*
+We dispatch live update events to notify the consuming app about added / updated posts.
+
+Consuming app uses these events to execute tasks like initialising Origami components
+on the updated elements.
+
+We want the rendering of the updates in the DOM to finish before dispatching this event,
+because the consumer needs to reference the updated DOM elements.
+
+If we dispatch the event in the same event loop with DOM element updates, consumer app
+will handle the event before the updates are complete.
+
+window.setTimeout(fn, 0) will defer the execution of the inner function until the
+current event loop completes, which is enough time for the DOM updates to finish.
+
+More information can be found in MDN setTimeout documentation. Please refer to
+"Late timeouts" heading in this page:
+https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/setTimeout#Late_timeouts
+
+> ... the timeout can also fire later when the page (or the OS/browser itself) is busy
+> with other tasks. One important case to note is that the function or code snippet
+> cannot be executed until the thread that called setTimeout() has terminated.
+> ...
+> This is because even though setTimeout was called with a delay of zero, it's placed on
+> a queue and scheduled to run at the next opportunity; not immediately.
+*/
+const dispatchEvent = (eventType, data) => {
+	if (typeof window === 'undefined') {
+		return
+	}
+	window.setTimeout(() => document.dispatchEvent(new CustomEvent(eventType, { detail: data })), 0)
```

Comment:
I think this will be ok but we need to documented way to target those events that is unique per instance. 

So either a unique event name or something within the `event.detail` that allows those events to be targeted to a single instance. I imagine this uniqueness would be the article UUID

Reply:
So I think we can see this function as a generic dispatcher and the uniqueness will come from the namespace `LiveBlogWrapper` and the event name `INSERT_POST` (`LiveBlogWrapper.INSERT_POST`). Does that make sense?

Reply:
Unfortunately thats not the level of uniqueness I am looking for, that only separates it from other events by other components. If there are multiple uses of `x-live-blog-wrapper` then we need to be able to distinguish that an event comes from a specific instance.

```js
import { LiveBlogWrapper } from '@financial-times/x-live-blog-wrapper';

const liveBlogWrapperProps1 = {} // Only added for illustrative purposes so just empty
const liveBlogWrapperProps2 = {} // Only added for illustrative purposes so just empty

<LiveBlogWrapper {...liveBlogWrapperProps1} />

<LiveBlogWrapper {...liveBlogWrapperProps2} />

```

So in something like this ^ we should be able to identify events from a specific instance (1 or 2)

Reply:
Ah I see, are we expecting this wrapper to be used multiple times on the page?

Reply:
So we might not right now but we should build it as if it was. This is where my thinking of "do as little with `document` as possible" comes from.

Kara put it better than that though and said the guidance is "don’t do things that would break if more than one instance of the component exists [...] which, doing things on the document is one thing that often causes that"

Reply:
So using `document` is not a no go but should be used in a way that allows individual targeting

Reply:
I think in this case it may be more of a pure argument than a practical one.
I can make this change but it puts me back in the tricky situation I was trying to send events to an individual wrapper.
It's probably possible but will require more work since the actions intentionally don't know anything about the instance itself.
I think pragmatically this is not necessary even though I agree with the general principle, since this is a single page level component and not something that is repeated.
It's up to you, will just take more time blocking release.

Reply:
Unfortunately it is a practical argument as its there to protect against bugs in the app which could end up rendering multiple wrappers due to its gallery view. 

What about if we did this.

1. [At this point](https://github.com/Financial-Times/x-dash/pull/576/files#diff-8da901f23570248e5f8c0c02d0939e5f02dc6499bd6f93254856f2f55b8be665R58) instead of only passing the `post` we also pass the [liveBlogPackageUuid](https://github.com/Financial-Times/x-dash/blob/22fe694711ec716eb33a3fa580e105c4de3ce704/components/x-live-blog-wrapper/src/LiveEventListener.js#L14)
2. Then the [`insertPost` action](https://github.com/Financial-Times/x-dash/blob/22fe694711ec716eb33a3fa580e105c4de3ce704/components/x-live-blog-wrapper/src/LiveBlogWrapper.jsx#L15) could use the `liveBlogPackageUuid` as part of the event name.

Reply:
>  in the app which could end up rendering multiple wrappers due to its gallery view.
I'd prefer if we knew this for sure but will concede regardless.

Reply:
I have an idea to try and use the reference that's in the props, that may work...

Reply:
> I'd prefer if we knew this for sure but will concede regardless


We sort of do know this, we know the app renders 3 articles at a time and we know we are no in control of stopping that being 2 live blogs

Reply:
At the moment we're not using these events in the app, so it's not clear what issues we might run into at this point. Would be easier to reason about this if we had an implementation to test.